src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs

```diff
@@ -207,51 +207,104 @@ private void SignalCompletion()
             }
             _completed = true;
 
-            if (_continuation != null || Interlocked.CompareExchange(ref _continuation, ManualResetValueTaskSourceCoreShared.s_sentinel, null) != null)
+            if (_continuation is null && Interlocked.CompareExchange(ref _continuation, ManualResetValueTaskSourceCoreShared.s_sentinel, null) is null)
             {
-                if (_executionContext != null)
+                return;
+            }
+
+            if (_executionContext is null)
+            {
+                if (_capturedContext is null)
+                {
+                    if (RunContinuationsAsynchronously)
+                    {
+                        ThreadPool.UnsafeQueueUserWorkItem(_continuation, _continuationState, preferLocal: true);
+                    }
+                    else
+                    {
+                        _continuation(_continuationState);
+                    }
+
+                    return;
+                }
+
+                InvokeSchedulerContinuation();
```

Comment:
What's the goal of factoring this out into a helper?  If it's to streamline the mainline case, seems other parts of this should also be moved out into helpers?  Or should it just be manually inlined and the helper deleted?  I'm just wondering about why two of the four cases are in helpers and the other two aren't.

Reply:
`InvokeInlineContinuation` is in a helper as it has error handling which makes the rest of the asm inefficient (causes stack spills etc); can inline the scheduler one

Reply:
Thanks.  Can you add a comment to that effect?

Reply:
Ah, `InvokeSchedulerContinuation` is called from two places, everything else is different which is why I kept that in a helper

Reply:
Added comment